### PR TITLE
fix(inputs.s7comm): Fix bit queries

### LINKED ISSUE #14043: inputs.s7comm bit parsing
### Relevant telegraf.conf

```toml
[[inputs.s7comm]]
  server = "10.0.0.1:102"
  rack = 0
  slot = 2

  ## Metric definition(s)
  [[inputs.s7comm.metric]]
    name = "Sinumerik"

    fields = [  
      {address= "DB1050.X0.1",    name= "RedGrindingActive"     },
      {address= "DB1050.X2.1",    name= "BlueGrindingActive"    },
  
      {address= "DB21.X35.5",       name= "NC running"           },
      {address= "DB21.X36.6",       name= "Channnel Alarm active"},
  
      {address= "PE32.X32.1",  name= "Controll On"      },
      {address= "PE34.X34.2",  name= "Emergency Stop"   },
      {address= "PE32.X32.7",  name= "Automatic active" },
      {address= "PE2.X2.7",   name= "KeyPos0-Auto"     },
      {address= "PE2.X2.6",   name= "KeyPos1"          },
      {address= "PE3.X3.6",   name= "KeyPos2"          },
      {address= "PE4.X4.4",   name= "KeyPos3-Setup"    },
      
    ]

    ## Tags assigned to the metric
    [inputs.s7comm.metric.tags]
      device = "Sinumerik"


[[outputs.discard]]
```

### Logs from Telegraf

```text
2023-10-03T09:28:28Z I! Loading config: telegraf.conf
2023-10-03T09:28:28Z I! Starting Telegraf 1.29.0-5eaabde9 brought to you by InfluxData the makers of InfluxDB
2023-10-03T09:28:28Z I! Available plugins: 15 inputs, 9 aggregators, 29 processors, 24 parsers, 4 outputs, 5 secret-stores
2023-10-03T09:28:28Z I! Loaded inputs: s7comm
2023-10-03T09:28:28Z I! Loaded aggregators: 
2023-10-03T09:28:28Z I! Loaded processors: 
2023-10-03T09:28:28Z I! Loaded secretstores: 
2023-10-03T09:28:28Z I! Loaded outputs: discard
2023-10-03T09:28:28Z I! Tags enabled: host=XXXXX
2023-10-03T09:28:28Z I! [agent] Config: Interval:10s, Quiet:false, Hostname:"XXXXX", Flush Interval:10s
2023-10-03T09:28:28Z D! [agent] Initializing plugins
2023-10-03T09:28:28Z D! [agent] Connecting outputs
2023-10-03T09:28:28Z D! [agent] Attempting connection to [outputs.discard]
2023-10-03T09:28:28Z D! [agent] Successfully connected to outputs.discard
2023-10-03T09:28:28Z D! [agent] Starting service inputs
2023-10-03T09:28:30Z D! [inputs.s7comm] Reading batch 1...
2023-10-03T09:28:30Z D! [inputs.s7comm]   got [1] for field "RedGrindingActive" @ 0 --> false (bool)
2023-10-03T09:28:30Z D! [inputs.s7comm]   got [0] for field "BlueGrindingActive" @ 2 --> false (bool)
2023-10-03T09:28:30Z D! [inputs.s7comm]   got [1] for field "NC running" @ 35 --> false (bool)
2023-10-03T09:28:30Z D! [inputs.s7comm]   got [1] for field "Channnel Alarm active" @ 36 --> false (bool)
2023-10-03T09:28:30Z D! [inputs.s7comm]   got [0] for field "Controll On" @ 32 --> false (bool)
2023-10-03T09:28:30Z D! [inputs.s7comm]   got [0] for field "Emergency Stop" @ 34 --> false (bool)
2023-10-03T09:28:30Z D! [inputs.s7comm]   got [0] for field "Automatic active" @ 32 --> false (bool)
2023-10-03T09:28:30Z D! [inputs.s7comm]   got [0] for field "KeyPos0-Auto" @ 2 --> false (bool)
2023-10-03T09:28:30Z D! [inputs.s7comm]   got [0] for field "KeyPos1" @ 2 --> false (bool)
2023-10-03T09:28:30Z D! [inputs.s7comm]   got [0] for field "KeyPos2" @ 3 --> false (bool)
2023-10-03T09:28:30Z D! [inputs.s7comm]   got [1] for field "KeyPos3-Setup" @ 4 --> false (bool)
^C2023-10-03T09:28:34Z D! [agent] Stopping service inputs
2023-10-03T09:28:34Z D! [agent] Input channel closed
2023-10-03T09:28:34Z I! [agent] Hang on, flushing any cached metrics before shutdown
2023-10-03T09:28:34Z D! [outputs.discard] Wrote batch of 1 metrics in 571ns
2023-10-03T09:28:34Z D! [outputs.discard] Buffer fullness: 0 / 10000 metrics
2023-10-03T09:28:34Z I! [agent] Stopping running outputs
2023-10-03T09:28:34Z D! [agent] Stopped Successfully
```

### System info

Telegraf 1.29, Ubuntu 22.04 LTS, S7 CPU 317-3PN/DP

### Docker

_No response_

### Steps to reproduce

1. Run Telegraf 1.29 with supplied config against S7 CPU 317-3PN/DP

### Expected behavior

Values to be parsed correctly as indicated in Debug log. See Actual behavior

### Actual behavior

```
got [1] for field "NC running" @ 35 --> false (bool)
```

Value being parsed to `false` independently of actual value indicate by byte `[1]`

### Additional info

Change parsing function in 'telegraf/plugins/inputs/s7comm/type_conversion.go' Line 16 to 
```	
case "X":
	return func(buf []byte) interface{} {
		return buf[0] != 0
	}
```
Might be a local problem. 
@srebhan Could you please explain the original function `(buf[0] & (1 << extra)) != 0` so I don't miss something?

### PR DESCRIPTION
**Pull request by phagemann**
- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves #14043 

Update Gos7 to allow query of bits. Enforce valid range (0-7).

### CODE REVIEW

**Review by Hipska — APPROVED** [review]

**Comment by Hipska** [comment]
Do you think it will take long before [robinson/gos7](https://github.com/robinson/gos7) will be patched?

**Comment by phagemann** [comment]
Working on a PR to fix this. 
Honestly i am not sure about the time schedule. Maybe a month. 

At least 2b34129c144ad72b739a6ab3ccd1bcd960527455 is necessary anyway.

File: plugins/inputs/s7comm/s7comm.go [hunk 1]
```diff
@@ -40,7 +40,7 @@ var (
 	}
 	// Word-length mapping taken from https://github.com/robinson/gos7/blob/master/client.go
 	wordLenMap = map[string]int{
-		"X":  0x01, // Bit
+		"X":  0x02, // Bit | Query as byte, as bit address can't be specified in S7DataItem
```
Comment (phagemann) [inline plugins/inputs/s7comm/s7comm.go:43]: This needs to be reverted.

**Comment by phagemann** [comment]
Both needs to be changed back to `0x01`:
https://github.com/influxdata/telegraf/blob/f4c56e1597956cd95eb37336e0c08ae361338aba/plugins/inputs/s7comm/s7comm.go#L43
https://github.com/influxdata/telegraf/blob/0c1e21398e08a82e8ab53b49d2c226f04b517c12/plugins/inputs/s7comm/s7comm_test.go#L260

Commit 30a5940728a900efc151839778494b63b41580a3 and c28e49b4ad3541043cdd00a7b87a1b1c60d88eaa would need to be reverted.

**Comment by Hipska** [comment]
Please mark that in the code. (If @srebhan agrees to that)

**Comment by srebhan** [comment]
@phagemann I would love to see this fixed upstream in `gos7` and then change this PR to change whatever is necessary (and update the gos7 version). The problem with temporary fixes is that nobody will remove them once they are in. ;-)

Can you fix this upstream? I've created some PRs there and usually they are quick in merging that stuff...

**Comment by phagemann** [comment]
That's fine for me. 

I will do the upstream fix and come back afterwards 😃.

**Comment by powersj** [comment]
@phagemann it looks like your upstream issue was closed due to your fix in https://github.com/robinson/gos7/pull/65

Are we good to update the version of s7comm instead of this PR now?

**Comment by Hipska** [comment]
@powersj he stated that the PR would still be needed:

> Working on a PR to fix this. Honestly i am not sure about the time schedule. Maybe a month.
> 
> At least [2b34129](https://github.com/influxdata/telegraf/commit/2b34129c144ad72b739a6ab3ccd1bcd960527455) is necessary anyway.

**Comment by powersj** [comment]
Then the title and commit message need to be updated, @phagemann needs to confirm my original question, and afterwards we can re-review

**Comment by phagemann** [comment]
@powersj In order to solve this, I need to adjust the MR. Just updating won't work. Shouldn't be much work, but I would like to test it on a real system. If everything works well, I will have one available by the end of the week. 

> @phagemann it looks like your upstream issue was closed due to your fix in [robinson/gos7#65](https://github.com/robinson/gos7/pull/65)
> 
> Are we good to update the version of s7comm instead of this PR now?

File: plugins/inputs/s7comm/s7comm.go (hunk 1 above)
Comment (srebhan) [inline plugins/inputs/s7comm/s7comm.go:43]: Do you still need this?

**Comment by GitTurboy** [comment]
The problem is still existed in current release

**Comment by Hipska** [comment]
@GitTurboy correct, this PR is still a work in progress and is not merged yet.

File: plugins/inputs/s7comm/s7comm.go (hunk 1 above)
Comment (phagemann) [reply plugins/inputs/s7comm/s7comm.go]: Nope, should be reverted.
2b34129c144ad72b739a6ab3ccd1bcd960527455 is necessary to address @GitTurboy 's Problem.

File: plugins/inputs/s7comm/s7comm.go [hunk 2]
```diff
@@ -348,6 +362,7 @@ func handleFieldAddress(address string) (*gos7.S7DataItem, converterFunc, error)
 	item := &gos7.S7DataItem{
 		Area:     area,
 		WordLen:  wordlen,
+		Bit:      extra,
```
Comment (phagemann) [inline plugins/inputs/s7comm/s7comm.go:365]: The field Bit will be ignored, unless `WordLen == s7wlbit` (=1). Including this doesn't hurt other queries, but feels like bad practice.
Alternatively the result for bit-queries could be returned after L333, but this obscures the progam flow, as the buffer is created afterwards.

Would be happy about some ideas, how to solve this more elegantly.

**Review by Hipska — APPROVED** [review]

File: plugins/inputs/s7comm/s7comm.go [hunk 3]
```diff
@@ -317,6 +317,20 @@ func handleFieldAddress(address string) (*gos7.S7DataItem, converterFunc, error)
...
+		if x == "" {
+			return nil, nil, errors.New("extra parameter required")
+		}
+
+		bit, err = strconv.Atoi(x)
+		if err != nil {
+			return nil, nil, fmt.Errorf("invalid extra parameter: %w", err)
+		} else if extra < 0 || extra > 7 {
+			// Ensure bit address is valid
+			return nil, nil, fmt.Errorf("invalid extra parameter %d, bit address out of range", extra)
```
Comment (Hipska) [inline plugins/inputs/s7comm/s7comm.go:332]: Typo?
```suggestion
		} else if bit < 0 || bit > 7 {
			// Ensure bit address is valid
			return nil, nil, fmt.Errorf("invalid extra parameter %d, bit address out of range", bit)
```

Comment (phagemann) [reply plugins/inputs/s7comm/s7comm.go]: Yes, thanks =)

File: plugins/inputs/s7comm/s7comm.go [hunk 4]
```diff
@@ -317,6 +317,20 @@ func handleFieldAddress(address string) (*gos7.S7DataItem, converterFunc, error)
...
+		if x == "" {
+			return nil, nil, errors.New("extra parameter required")
+		}
+
+		bit, err = strconv.Atoi(x)
+		if err != nil {
+			return nil, nil, fmt.Errorf("invalid extra parameter: %w", err)
+		} else if bit < 0 || bit > 7 {
+			// Ensure bit address is valid
+			return nil, nil, fmt.Errorf("invalid extra parameter %d, bit address out of range", bit)
```
Comment (Hipska) [inline plugins/inputs/s7comm/s7comm.go:332]: What do you think of this?
```suggestion
			return nil, nil, fmt.Errorf("invalid extra parameter: bit address %d out of range", bit)
```

**Review by Hipska — APPROVED** [review]

File: plugins/inputs/s7comm/s7comm.go [hunk 5]
```diff
@@ -317,6 +317,20 @@ func handleFieldAddress(address string) (*gos7.S7DataItem, converterFunc, error)
...
+			return nil, nil, errors.New("extra parameter required")
+		}
+
+		bit, err = strconv.Atoi(x)
+		if err != nil {
+			return nil, nil, fmt.Errorf("invalid extra parameter: %w", err)
+		} else if bit < 0 || bit > 7 {
+			// Ensure bit address is valid
+			return nil, nil, fmt.Errorf("invalid extra parameter: bit address %d out of range", bit)
+		}
```
Comment (srebhan) [inline plugins/inputs/s7comm/s7comm.go:333]: We don't need the else as we do return in the `if`...
```suggestion
		}
		if bit < 0 || bit > 7 {
			// Ensure bit address is valid
			return nil, nil, fmt.Errorf("invalid extra parameter: bit address %d out of range", bit)
		}
```

File: plugins/inputs/s7comm/type_conversions.go [hunk 6]
```diff
@@ -9,11 +9,11 @@ import (
 
 var helper = &gos7.Helper{}
 
-func determineConversion(dtype string, extra int) converterFunc {
+func determineConversion(dtype string, _ int) converterFunc {
```
Comment (srebhan) [inline plugins/inputs/s7comm/type_conversions.go:12]: Please remove the parameter altogether if it is not used...

**Comment by GitTurboy** [comment]
> Download PR build artifacts for [linux_amd64.tar.gz](https://output.circle-artifacts.com/output/job/179de5c8-3d7a-4ff5-939d-a3da819b1c7f/artifacts/0/build/dist/telegraf-1.29.0~2b009478_linux_amd64.tar.gz), [darwin_amd64.tar.gz](https://output.circle-artifacts.com/output/job/ab8dfd20-b8c2-4543-a26f-5a11d85a950e/artifacts/0/build/dist/telegraf-1.29.0~2b009478_darwin_amd64.tar.gz), and [windows_amd64.zip](https://output.circle-artifacts.com/output/job/9ecca973-cb17-4f14-8e55-2331f5b4b796/artifacts/0/build/dist/telegraf-1.29.0~2b009478_windows_amd64.zip). Downloads for additional architectures and packages are available below.
> 
> 👍 This pull request doesn't change the Telegraf binary size
> 
> 📦 Click here to get additional PR build artifacts

Thank you all, but I found the 【Allow PDU-size to be set as config option】 problem（ https://github.com/influxdata/telegraf/pull/14045） were not included

**Review by srebhan — COMMENTED** [review]
@phagemann looks good overall, only two minor comments from my side...

File: plugins/inputs/s7comm/s7comm.go (hunk 5 above)
Comment (phagemann) [reply plugins/inputs/s7comm/s7comm.go]: Thanks for your feedback =) I appreciate it.

**Comment by GitTurboy** [comment]
I tried to use the last PR(telegraf-1.29.0~92c29c75_windows_amd64)  to fetch about 90 items  data from a smart 200 PLC, the log shown error message as below:
[inputs.s7comm] Error in plugin: reading batch 1 failed: CLI : too may items (>20) in multi read/write

**Comment by phagemann** [comment]
You should not exceed a pdu-size of 20. This is the maximum number of elements contained in a request batch. It does not limit the number of total items. Please try to set set it back to 20 or lower. 


> I tried to use the last PR(telegraf-1.29.0~92c29c75_windows_amd64) to fetch about 90 items data from a smart 200 PLC, the log shown error message as below:[ I set pdu_size = 200] [inputs.s7comm] Error in plugin: reading batch 1 failed: CLI : too may items (>20) in multi read/write

**Comment by GitTurboy** [comment]
> The error message is produced by Gos7. You should not exceed a pdu-size of 20. This is the maximum number of elements contained in a request batch. It does not limit the number of total items. Please try to set set it back to 20 or lower.
> 
> > I tried to use the last PR(telegraf-1.29.0~92c29c75_windows_amd64) to fetch about 90 items data from a smart 200 PLC, the log shown error message as below:[ I set pdu_size = 200] [inputs.s7comm] Error in plugin: reading batch 1 failed: CLI : too may items (>20) in multi read/write

But when I set pdu-size to 20 , the message in  log will shown:

2023-11-13T00:19:27Z D! [inputs.s7comm] Reading batch 1...
2023-11-13T00:19:27Z E! [inputs.s7comm] Error in plugin: reading batch 1 failed: CPU : total data exceeds the PDU size

**Comment by phagemann** [comment]
Now your CPU (smart 200) complains about the pdu-size beeing too large. Try lowering it, until the CPU stops reporting this error.

Not sure, whether this is this the right place, as it referring to a  already closed issue. It is not relevant for this MR.

**Comment by GitTurboy** [comment]
> Now your CPU (smart 200) complains about the pdu-size beeing too large. Try lowering it, until the CPU stops reporting this error.
> 
> Not sure, whether this is this the right place, as it referring to a already closed issue. It is not relevant for this MR.

Thanks， this method resolved  the problem.

**Review by Hipska — APPROVED** [review]

**Review by srebhan — APPROVED** [review]
Thanks for fixing this @phagemann!

**Review by powersj — APPROVED** [review]

**Pull request merged**